### PR TITLE
Neutral positivity

### PR DESCRIPTION
**Pull request by tnbernard**
Request to merge this branch which adds option for nonconservative positivity preserving scheme. This scheme finds cells with cell averages less than zero and sets all coefficients to zero. This is motivated by the negativity arising from steep density gradients in neutral species profiles due to the wall recycling boundary conditions. Nonconservative positivity can be called from within the species table in the Lua input file by including the following line.

`nonconPositivity = true,`

All regression tests that pass on master pass on this branch. However, gk-neutrals will need to be regenerated due to the following: (1) intSrcIzM0 is no longer integrated in time, affecting all tests, (2) BCs in rt-nstx-lowRes-p1.lua now include recycling BCs and the correct radial Vlasov BCs for neutrals, and (3) the addition of a new rt for nonconPositivity. 

Other positivity schemes in this branch should not be affected whatsoever by these changes.

### CODE REVIEW

File: App/Species/KineticSpecies.lua [hunk 1]
```diff
@@ -220,6 +220,7 @@ function KineticSpecies:fullInit(appTbl)
    self.positivity        = xsys.pickBool(tbl.applyPositivity, false)
    self.positivityDiffuse = xsys.pickBool(tbl.positivityDiffuse, self.positivity)
    self.positivityRescale = xsys.pickBool(tbl.positivityRescale, false)
+   self.nonconPositivity  = xsys.pickBool(tbl.nonconPositivity, false)
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:223]: Are `positivity` and `nonconPositivity` exclusive? If yes maybe there should be a check to make sure use doesn't request both?

File: App/Species/KineticSpecies.lua [hunk 2]
```diff
@@ -466,6 +467,25 @@ function KineticSpecies:createSolver(field, externalField)
...
+      self.postPosM0    = self:allocMoment()
+      self.delPosM0     = self:allocMoment()
+      self.intDelPosM0  = DataStruct.DynVector{numComponents = 1}
+      self.intPosM0     = DataStruct.DynVector{numComponents = 1}
+      self.calcIntPosM0 = Updater.CartFieldIntegratedQuantCalc {
+	    onGrid        = self.confGrid,
+	    basis         = self.confBasis,
+	    numComponents = 1,
+	    quantity      = "V",
+	    timeIntegrate = timeIntegrate,
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:481]: Should this line be removed? I don't see where the `timeIntegrate` on the right is defined.

File: App/Species/KineticSpecies.lua [hunk 3]
```diff
@@ -466,6 +467,25 @@ function KineticSpecies:createSolver(field, externalField)
...
+      }
+      self.nonconPos = Updater.PositivityRescale {
+         onGrid = self.grid,
+         basis  = self.basis,
+	 nonconservative = true,
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:486]: No tabs would be preferable (if any).

File: App/Species/KineticSpecies.lua [hunk 4]
```diff
@@ -685,7 +705,7 @@ end
 function KineticSpecies:checkPositivity(tCurr, idx)
   local status = true
   if self.positivity then
-     status = self.posChecker:advance(tCurr, {self:rkStepperFields()[idx]}, {})
+     --status = self.posChecker:advance(tCurr, {self:rkStepperFields()[idx]}, {})
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:708]: Is this meant to stay commented out? if so, why not remove it?

File: App/Species/KineticSpecies.lua [hunk 5]
```diff
@@ -700,6 +720,19 @@ function KineticSpecies:applyBcIdx(tCurr, field, externalField, inIdx, outIdx, i
...
+     self.nonconPos:advance(tCurr, {self:rkStepperFields()[outIdx]}, {self:rkStepperFields()[outIdx]}, false)
+     
+     self.numDensityCalc:advance(tCurr, {self:rkStepperFields()[outIdx]}, {self.postPosM0})
+     self.delPosM0:combine(1.0, self.postPosM0, -1.0, self.prePosM0)
+
+     local tm, lv = self.intPosM0:lastData()
+     self.calcIntPosM0:advance(tCurr, {self.delPosM0}, {self.intDelPosM0})
+     local tmDel, lvDel = self.intDelPosM0:lastData()
+     self.intPosM0:appendData(tmDel, {lv[1]+lvDel[1]})
+  end
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:735]: I think we shouldn't add more branching in code that gets called inside the time loop. So I would advocate for/request that the code instead checks the positivity booleans and creates appropriate functions during initialization. When the booleans are false those functions would be empty. See App/Species/FluidSpecies.lua, which is probably the most advanced in this regard. Its :applyBcIdx method simply calls a function self.checkPositivity whose definition depends on the booleans in the input file. If `positivity` and `nonconPositivity` are exclusive, then you could create a single function that does either.
The hope is that such approach reduces overhead. The compiler either simply inlines the function or in the case of empty functions, removes it altogether.

File: App/Species/KineticSpecies.lua [hunk 6]
```diff
@@ -778,6 +811,11 @@ function KineticSpecies:calcAndWriteDiagnosticMoments(tm)
        self.vSigmaCX:write(string.format("%s_vSigmaCX_%d.bp", self.name, self.diagIoFrame), tm, self.diagIoFrame, self.writeSkin)
        self.collisions[self.collNmCX].sourceCX:write(string.format("%s_sourceCX_%d.bp", self.name, self.diagIoFrame), tm, self.diagIoFrame, self.writeSkin)
     end
+
+    -- Vlasov positivity diagnostics
+    if self.nonconPositivity then
+       self.intPosM0:write( string.format("%s_intPosM0.bp", self.name), tm, self.diagIoFrame)
+    end
```
Comment (manauref) [inline App/Species/KineticSpecies.lua:818]: Can you do this through the Diagnostics App infrastructure? as commented at the beginning of this method, :calcAndWriteDiagnosticMoments will hopefully disappear.

**Review by manauref — COMMENTED** [review]
Aside from my comment on `timeIntegrate` and a line that appears to be commented out, I don't see any obvious bugs or errors. I do however advocate for
1. Not adding more branching inside the time loop and instead creating appropriate functions during initialization (see FluidSpecies).
2. Using the Diagnostics App infrastructure instead of adding to methods that will hopefully disappear.

File: App/Species/KineticSpecies.lua (hunk 5 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua:735]: Sounds good. Will do.

File: App/Species/KineticSpecies.lua (hunk 6 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua:818]: Yes, I can make those changes.

File: App/Species/KineticSpecies.lua (hunk 4 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua:708]: Thanks for catching that. I meant to delete.

File: App/Species/KineticSpecies.lua (hunk 1 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua:223]: Will think about how to do this appropriately.

File: App/Species/KineticSpecies.lua (hunk 2 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua:481]: Also meant to delete that. Thanks for catching.

**Comment by tnbernard** [comment]
It will take some time to resolve these comments and test the results. I will push in a day or two once completed.

File: App/Species/KineticSpecies.lua (hunk 5 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua]: Commit c7b8dcd addresses this.

File: App/Species/KineticSpecies.lua (hunk 6 above)
Comment (tnbernard) [reply App/Species/KineticSpecies.lua]: :calcAndWriteDiagnosticMoments has been deleted. Now the write function is done as in FluidSpecies for positivity.
